animate/metric.py

```diff
@@ -259,12 +291,88 @@ def enforce_spd(self, restrict_sizes=False, restrict_anisotropy=False):
             "restrictSizes": restrict_sizes,
             "restrictAnisotropy": restrict_anisotropy,
         }
+        if self._variable_parameters_set:
+            kw["restrictSizes"] = False
+            kw["restrictAnisotropy"] = False
         v = self._create_from_array(self.dat.data_with_halos)
         det, _ = self._plex.metricDeterminantCreate()
         self._plex.metricEnforceSPD(v, v, det, **kw)
         size = np.shape(self.dat.data_with_halos)
         self.dat.data_with_halos[:] = np.reshape(v.array, size)
         v.destroy()
+        if self._variable_parameters_set:
+            if restrict_sizes and restrict_anisotropy:
+                return self._enforce_variable_constraints()
+            elif restrict_sizes or restrict_anisotropy:
+                raise NotImplementedError(
+                    "Can only currently restrict both sizes and anisotropy."
+                )
+        return self
+
+    # TODO: Implement this on the PETSc side
+    #       See https://gitlab.com/petsc/petsc/-/issues/1450
+    @PETSc.Log.EventDecorator()
+    def _enforce_variable_constraints(self):
+        """
+        Post-process a metric to enforce minimum and maximum metric magnitudes
+        and maximum anisotropy, any of which may vary spatially.
+        """
+        mesh = self.function_space().mesh()
+        P1 = firedrake.FunctionSpace(mesh, "CG", 1)
+
+        def interp(f):
+            r"""
+            Try to apply a Clement interpolant.
+              * `TypeError` indicates something other than a :class:`Function` passed.
+              * `ValueError` indicates the :class:`Function` is not :math:`\mathbb{P}0`.
+            """
+            try:
+                return clement_interpolant(f)
```

Comment:
Let's say we want the same metric parameters on each subinterval metric. If the constraints are not variable, we can just pass a single metric params dictionary to `goalie.metric.space_time_normalise` which then makes a list of the required length (=number of subintervals). I think this is quite nice. However, if we have variable constraints we always have to pass a list of metric params dictionaries, and we have to modify each respective dictionary. I.e., something like this:

```
    mp1, mp2 = {...}, {...}
    mps = [mp1, mp2]

    for i, mesh in enumerate(mesh_seq):
        P1 = FunctionSpace(mesh, "CG", 1)
        expression = ...
        mps[i]["dm_plex_metric"]["a_max"] = interpolate(expression, P1)
```

I think we can avoid this if we pass `target_space=P1` here. Otherwise there is a function spaces mismatch at L352 below.

Reply:
Sorry, I'm not sure I follow. What function/method is it that you are suggesting `target_space=P1` to? I might be misunderstanding you, but isn't what you're saying all about how the dictionaries are set up, as opposed to how the new `_enforce_variable_constraints` method works?

Reply:
@ddundo Just tagging you here in case you missed the notification.

Reply:
Ah sorry, I did! And sorry for the bad explanation. I don't want to change how dictionaries are set up.

If we want to set the same metric parameters on each subinterval, in the non-variable constraints case we can just pass a single dictionary of metric parameters. In the variable constraints case (here), we cannot - we have to pass a list of `len(mesh_sequence)` dictionaries, so that these variable constraints functions are defined over appropriate function spaces/meshes. I think we can avoid this if we pass target space to `clement_interpolant(f, target_space=P1)`? 

Again, please tell me if this is a bad idea / bad practice! And either way not too important - so feel free to ignore :)

Reply:
Ah apologies, I had been confused which function/method you were suggesting to pass the `target_space` kwarg to - this makes sense. Addressed in 3533e8b.